main/main.c

```diff
@@ -659,7 +659,40 @@ PHPAPI ZEND_COLD void php_log_err_with_severity(char *log_message, int syslog_ty
 	if (PG(error_log) != NULL) {
 #ifdef HAVE_SYSLOG_H
 		if (!strcmp(PG(error_log), "syslog")) {
-			php_syslog(syslog_type_int, "%s", log_message);
+			static char nvt[] = " !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~";
+			char *nl, *msg;
+			for (msg = log_message; msg != NULL; msg = nl + 1) {
+				size_t valid_len;
+
+				/* if the message contains invalid characters,
+				 * rewrite it; otherwise print it as is
+				 */
+				valid_len = strspn(msg, nvt);
+
+				if (msg[valid_len] == '\0' || msg[valid_len] == '\n') {
+					php_syslog(syslog_type_int, "%.*s", (int)valid_len, msg);
+					if (msg[valid_len] == '\0')
+						break;
+					nl = &msg[valid_len];
+				} else {
+					char *buf, *ptr;
+					size_t buflen;
+
+					FILE *fp = open_memstream(&buf, &buflen);
```

Comment:
By the way, I have to back-port this patch to 7.1.8, which doesn't have `zend_printf_to_smart_string()`... what's the workaround?